### PR TITLE
Sort comments reopen

### PR DESCRIPTION
**Pull request by vishaldeepak**
## What type of PR is this? (check all applicable)

- [* ] Feature

## Description

Add the ability the sort comments of Articles View Page. This feature was already merged earlier , [here](https://github.com/forem/forem/pull/16686).  But after deployment issues were found and it was [reverted](https://github.com/forem/forem/pull/17093).  I've followed advice from Aitchiss I've resolved all issues described in this [comment](https://github.com/forem/forem/issues/16133#issuecomment-1205094470). 

## Related Tickets & Documents

- Related Issue #
  - https://github.com/forem/forem/issues/16133
- Closes #
  - https://github.com/forem/forem/issues/17084 
  - https://github.com/forem/forem/issues/17082
  - https://github.com/forem/forem/issues/17081

## QA Instructions, Screenshots, Recordings

Any article page with comments should now show the ability sort comments in three ways 
- Top
- Latest 
- Oldest

Web View

Screenshot 2022-08-18 at 12 13 01 AM

Mobile View

Screenshot 2022-08-18 at 12 13 38 AM

## Added/updated tests?

- [*] Yes
- [ ] No, and this is why: Caching Edge Case dont have test cases

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://developers.forem.com) or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [ ] I've updated the README or added inline documentation
- [ ] I've added an entry to
      [`CHANGELOG.md`](https://github.com/forem/forem/tree/main/CHANGELOG.md)
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [ ] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] What gif best describes this PR or how it makes you feel?

alt_text

### CODE REVIEW

**Comment by vishaldeepak** [comment]
There's one concern in this PR. I randomly get `{"error":"Error: Bad Request"}`. Its hard to debug this since logs dont mention what went wrong.

```
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0] Started GET "/abshirebrian/little-hands-clapping-voluptatem-ut-83n" for 127.0.0.1 at 2022-08-18 00:17:12 +0530
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]    (0.6ms)  SELECT "flipper_features"."key" AS feature_key, "flipper_gates"."key", "flipper_gates"."value" FROM "flipper_features" LEFT OUTER JOIN "flipper_gates" ON "flipper_features"."key" = "flipper_gates"."feature_key"
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0] Processing by StoriesController#show as HTML
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Parameters: {"locale"=>nil, "username"=>"abshirebrian", "slug"=>"little-hands-clapping-voluptatem-ut-83n"}
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["id", 11], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/controllers/concerns/edge_cache_safety_check.rb:12:in `current_user'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Settings::SMTP Load (0.2ms)  SELECT "settings_smtp"."var", "settings_smtp"."value" FROM "settings_smtp"
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/settings/base.rb:147:in `each_with_object'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Settings::General Load (0.2ms)  SELECT "site_configs"."var", "site_configs"."value" FROM "site_configs"
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/settings/base.rb:147:in `each_with_object'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Settings::UserExperience Load (0.2ms)  SELECT "settings_user_experiences"."var", "settings_user_experiences"."value" FROM "settings_user_experiences"
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/settings/base.rb:147:in `each_with_object'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Article Load (0.5ms)  SELECT "articles".* FROM "articles" WHERE "articles"."path" = $1 LIMIT $2  [["path", "/abshirebrian/little-hands-clapping-voluptatem-ut-83n"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/controllers/stories_controller.rb:37:in `show'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1  [["id", 10]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/controllers/stories_controller.rb:37:in `show'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Settings::General Load (0.3ms)  SELECT "site_configs".* FROM "site_configs" WHERE "site_configs"."var" = $1 LIMIT $2  [["var", "feed_pinned_article_id"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/pinned_article.rb:44:in `setting'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   DiscussionLock Load (0.2ms)  SELECT "discussion_locks".* FROM "discussion_locks" WHERE "discussion_locks"."article_id" = $1 LIMIT $2  [["article_id", 22], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/controllers/stories_controller.rb:255:in `assign_article_show_variables'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Settings::Community Load (0.3ms)  SELECT "settings_communities"."var", "settings_communities"."value" FROM "settings_communities"
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/settings/base.rb:147:in `each_with_object'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendering layout layouts/application.html.erb
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendering articles/show.html.erb within layouts/application
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered shared/_payment_pointer.html.erb (Duration: 0.0ms | Allocations: 44)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered shared/_webcomponents_loader_script.html.erb (Duration: 0.0ms | Allocations: 43)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_reaction_button.html.erb (Duration: 1.9ms | Allocations: 3498)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_reaction_button.html.erb (Duration: 1.7ms | Allocations: 3485)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_reaction_button.html.erb (Duration: 1.8ms | Allocations: 3485)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_actions.html.erb (Duration: 9.8ms | Allocations: 17880)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   CACHE Settings::General Load (0.0ms)  SELECT "site_configs".* FROM "site_configs" WHERE "site_configs"."var" = $1 LIMIT $2  [["var", "feed_pinned_article_id"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/pinned_article.rb:44:in `setting'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Tag Load (0.5ms)  SELECT "tags"."bg_color_hex", "tags"."text_color_hex" FROM "tags" WHERE "tags"."name" = $1 LIMIT $2  [["name", "discuss"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/helpers/application_helper.rb:154:in `block in tag_colors'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Tag Load (0.2ms)  SELECT "tags"."bg_color_hex", "tags"."text_color_hex" FROM "tags" WHERE "tags"."name" = $1 LIMIT $2  [["name", "linux"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/helpers/application_helper.rb:154:in `block in tag_colors'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Tag Load (0.2ms)  SELECT "tags"."bg_color_hex", "tags"."text_color_hex" FROM "tags" WHERE "tags"."name" = $1 LIMIT $2  [["name", "go"], ["LIMIT", 1]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/helpers/application_helper.rb:154:in `block in tag_colors'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_sort_option.html.erb (Duration: 1.0ms | Allocations: 1891)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_sort_option.html.erb (Duration: 0.0ms | Allocations: 51)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_sort_option.html.erb (Duration: 0.0ms | Allocations: 51)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_form.html.erb (Duration: 2.1ms | Allocations: 4520)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Comment Load (0.8ms)  SELECT "comments".* FROM "comments" WHERE "comments"."commentable_id" = $1 AND "comments"."commentable_type" = $2 ORDER BY score DESC  [["commentable_id", 22], ["commentable_type", "Article"]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/comment.rb:95:in `tree_for'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   User Load (0.3ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1  [["id", 6]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/comment.rb:95:in `tree_for'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Users::Setting Load (0.3ms)  SELECT "users_settings".* FROM "users_settings" WHERE "users_settings"."user_id" = $1  [["user_id", 6]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/comment.rb:95:in `tree_for'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Profile Load (0.3ms)  SELECT "profiles".* FROM "profiles" WHERE "profiles"."user_id" = $1  [["user_id", 6]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/comment.rb:95:in `tree_for'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_avatar.html.erb (Duration: 0.4ms | Allocations: 561)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_quality_marker.html.erb (Duration: 0.1ms | Allocations: 45)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ProfileField Pluck (0.3ms)  SELECT "profile_fields"."label", "profile_fields"."attribute_name" FROM "profile_fields" WHERE "profile_fields"."display_area" = $1  [["display_area", 0]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/decorators/profile_decorator.rb:5:in `ui_attributes_for'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered shared/_profile_card_content.html.erb (Duration: 18.2ms | Allocations: 1756)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered shared/_profile_preview_card.html.erb (Duration: 18.7ms | Allocations: 2512)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_date.erb (Duration: 0.4ms | Allocations: 489)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_header.html.erb (Duration: 22.1ms | Allocations: 8255)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_footer.html.erb (Duration: 2.8ms | Allocations: 5715)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment_proper.html.erb (Duration: 27.9ms | Allocations: 15218)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_comment.html.erb (Duration: 30.1ms | Allocations: 19129)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered collection of articles/_comment_tree.html.erb [1 times] (Duration: 30.3ms | Allocations: 19328)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_conduct_and_abuse_actions.html.erb (Duration: 0.1ms | Allocations: 121)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_comments_actions.html.erb (Duration: 0.2ms | Allocations: 316)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered comments/_hide_comments_modal.html.erb (Duration: 0.5ms | Allocations: 1315)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_full_comment_area.html.erb (Duration: 41.9ms | Allocations: 35446)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   DisplayAd Load (0.4ms)  SELECT "display_ads".* FROM "display_ads" WHERE "display_ads"."approved" = $1 AND "display_ads"."published" = $2 AND "display_ads"."placement_area" = $3 ORDER BY "display_ads"."success_rate" DESC LIMIT $4  [["approved", true], ["published", true], ["placement_area", "post_comments"], ["LIMIT", 3]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/display_ad.rb:29:in `for_display'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]    (0.6ms)  SELECT ( (reltuples / GREATEST(relpages, 1)) * (pg_relation_size('articles') / (GREATEST(current_setting('block_size')::integer, 1))) )::bigint AS count FROM pg_class WHERE relname = 'articles' AND relkind = 'r';
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/models/application_record.rb:22:in `estimated_count'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Article Load (3.3ms)  SELECT "articles".* FROM "articles" WHERE "articles"."published" = $1 AND (published_at <= '2022-08-17 18:47:12.432990') AND (cached_tag_list ~ '[[:<:]]discuss[[:>:]]' OR cached_tag_list ~ '[[:<:]]linux[[:>:]]' OR cached_tag_list ~ '[[:<:]]go[[:>:]]') AND "articles"."user_id" != $2 AND (score > 1) ORDER BY "articles"."hotness_score" DESC LIMIT $3 OFFSET $4  [["published", true], ["user_id", 10], ["LIMIT", 4], ["OFFSET", 2]]
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   ↳ app/services/articles/suggest.rb:56:in `suggestions_by_tag'
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/_bottom_content.html.erb (Duration: 0.7ms | Allocations: 1998)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered articles/show.html.erb within layouts/application (Duration: 67.7ms | Allocations: 69876)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0]   Rendered layout layouts/application.html.erb (Duration: 67.9ms | Allocations: 70067)
00:17:12 web.1       | [dd.env=development dd.service=rails dd.trace_id=76836365835559263 dd.span_id=0] Completed 400 Bad Request in 81ms (Views: 0.1ms | ActiveRecord: 9.7ms | Allocations: 83712)
00:17:12 web.1       |
```

File: app/views/comments/_sort_option.html.erb [hunk 1]
```diff
@@ -0,0 +1,15 @@
+<li class="comment-sort-option">
+  <a
+    href="<%= @article.path %>?comments_sort=<%= sort_order %>"
```
Comment (fdocr) [inline app/views/comments/_sort_option.html.erb:3]: I'm not sure if this was discussed at some point, but after selecting a filter the page is refreshed and users have to scroll back down to the comments section. Maybe we can add a simple anchor to the link to avoid this, what do you think?

```suggestion
    href="<%= @article.path %>?comments_sort=<%= sort_order %>#comments"
```

**Review by fdocr — COMMENTED** [review]
Hi @vishaldeepak, thank you for the contribution and your patience. This looks really nice!

I looked into the error you were facing and it seems like the errors were just a flaky spec that was fixed by rerunning the CI jobs.

I left a small inline suggestion for you, but after a quick look through the PR it seems like everything looks good to open up for review. Don't hesitate to ping me if you need any more help with this!

File: app/views/comments/_sort_option.html.erb (hunk 1 above)
Comment (vishaldeepak) [reply app/views/comments/_sort_option.html.erb:3]: Sure.  Any idea about error I posted https://github.com/forem/forem/pull/18349#issuecomment-1218378312

Comment (fdocr) [reply app/views/comments/_sort_option.html.erb:3]: So I've seen that before a few times but only in development mode (locally), but it seems like it happens when to try to access a post and the post itself is somewhat bugged. I believe this has happened after switching branches a few times and something has changed/modified over time.

My suggestion would be to merge main so your branch is up to date and then recreate the DB, hopefully you won't run into this anymore. Something along the lines of:

```ruby
# Drop the DB
bundle exec rails db:drop

# Recreate & seed the DB
bin/setup

#Note: Only if the local DB isn't seeded you can try to seed it manually with `bundle exec rails db:seed`

# Startup again
bin/startup
```

Let me know if this solves the issue you're seeing

Comment (vishaldeepak) [reply app/views/comments/_sort_option.html.erb]: seems to resolved, thank you

Comment (vishaldeepak) [reply app/views/comments/_sort_option.html.erb]: @fdocr Opened for Review!!! Excited

**Review by fdocr — APPROVED** [review]
Tested locally and LGTM!

Thank you @vishaldeepak for working on this feature 🙌🏼

**Review by jaw6 — APPROVED** [review]
:+1: Looks good to me!
